### PR TITLE
Make Healthcheck and Keepalive configurable by IP

### LINKED ISSUE #84: ETCD Keepalive time should be configurable by IP (potentially disablable)
Context :

Each IP is being keepalived with ETCD by maintaining a lease. For single-node IP, it's useless and it creates a lot of load on the ETCD clusters.

Action to take:

For IP with only 1 node, we don't need to ping every N seconds ETCD, it should be configurable by IP in order to make it a very long value, or to even disable keepalive lease

### PR DESCRIPTION
**Pull request by Soulou**
Fixes #84

### CODE REVIEW

File: api/http_client.go [hunk 1]
```diff
@@ -147,11 +147,20 @@ func (c HTTPClient) TryGetLock(ctx context.Context, id string) error {
 	return nil
 }
 
-func (c HTTPClient) AddIP(ctx context.Context, ip string, checks ...models.Healthcheck) (IP, error) {
+type AddIPParams struct {
+	IP                  string               `json:"ip"`
```
Comment (EtienneM) [inline api/http_client.go]: The IP argument being mandatory, I would prefer the following signature for this function:

```go
 func (c HTTPClient) AddIP(ctx context.Context, ip string, params AddIPParams) (IP, error) { 
```
What do you think?

File: ip/event_manager.go [hunk 2]
```diff
@@ -75,6 +75,10 @@ func (m *manager) eventManager(ctx context.Context) {
 			return
 		}
 
+		interval := time.Duration(m.IP().KeepaliveInterval) * time.Second
```
Comment (EtienneM) [inline ip/event_manager.go]: It's not a big computation, but don't you want to compute the interval outside of the loop?

File: ip/event_manager.go [hunk 3]
```diff
@@ -75,6 +75,10 @@ func (m *manager) eventManager(ctx context.Context) {
...
+		if interval == 0 {
+			interval = m.config.KeepAliveInterval
+		}
 		time.Sleep(m.config.KeepAliveInterval)
```
Comment (EtienneM) [inline ip/event_manager.go]:
```suggestion
		time.Sleep(interval)
```

File: ip/event_manager.go [hunk 4]
```diff
@@ -206,7 +210,11 @@ func (m *manager) healthChecker(ctx context.Context) {
 			m.sendHealthcheckResults(ctx, healthy, err)
 		}
 
-		time.Sleep(m.config.HealthcheckInterval)
+		interval := time.Duration(m.IP().HealthcheckInterval) * time.Second
```
Comment (EtienneM) [inline ip/event_manager.go]: It's not a big computation, but don't you want to compute the interval outside of the loop?

Comment (Soulou) [reply ip/event_manager.go]: ok

File: ip/event_manager.go (hunk 3 above)
Comment (Soulou) [reply ip/event_manager.go]: oups

File: ip/event_manager.go (hunk 2 above)
Comment (Soulou) [reply ip/event_manager.go]: ok

File: api/http_client.go (hunk 1 above)
Comment (Soulou) [reply api/http_client.go]: I agree

**Comment by johnsudaar** [comment]
This feature is nice, but i don't think it solve the issue.

The issue is that we make requests to etcd when there is only one host having the IP (so when it's not needed). In that case it should be link responsibility to manage that correctly and we should not rely on the client to do that.

**Comment by Soulou** [comment]
How would you build this feature? It would require much more changes no? Like registering the different instances competing for a given IP (so far this info is not available at all is that right?

Currently my problem is mostly understanding why do we need to wait twice the lease when removing an IP.  As we're doing 

```
* Unlock() (DELETE /etcd/key/path/IP)
* Wait until someone take the IP, or 2 × Lease Time 
* RemoveIP()
* Stopping manager and removing IP from storage
```

Why are we waiting that long ? Wouldn't it work to say 

```
* Unlock()
* Wait until someone take
* Wait for someone to take the IP or 1 minute
* RemoveIP
* RemoveStorage
```

**Comment by johnsudaar** [comment]
Indeed nodes does not know which other nodes are trying to get the same IP. But that is pretty straight forward to add.

For your question, we need to wait at least 1 Lease Time for someone else to take the IP. If not, we could release the IP before any other nodes try to get it. If we do so, the IP won't be assigned to any node which will lead to downtime.

If we wait only 1 LeaseTime, we might have some tricky race conditions that's why we choosed 2 LeaseTime to be extra safe.

**Comment by johnsudaar** [comment]
We could register all nodes in a /ips/[ID]/nodes/[NODE_ID] key and make all nodes watch the /ips/[ID] folder. So they could dynamically adapt their lease time on their own.

**Comment by johnsudaar** [comment]
(or even on the /ips/ prefix if you want to reduce connections)

**Comment by Soulou** [comment]
I'm a bit afraid of the "pretty straight forward to add" actually :D

That's an interesting way to handle the problem, I like it quite well, so we prefer to having a granular configuration?

But we would have more or less the same set of problems :

* sand is detecting it's alone to control an IP
* keepalive time becomes high
* link ip is removed (by database-core request)
* what should we wait?

For the second part the 2 lease times is much clearer to me, it's just that it doesn't seem to be a model which works for long leases.

**Comment by johnsudaar** [comment]
Currently LinK does not know if he's alone controlling the IP. So it assume it is in the worse case scenario (it's not alone). In this case, LinK should not remove the IP directly. It should wait for another node to take the IP. To do so it will not renew the lease and wait for someone else to take it.

In the worse case this step can take 2*LeaseTime (1 for the lease to expire, 1 for another one to take it).

**Comment by Soulou** [comment]
> Currently LinK does not know if he's alone controlling the IP. So it assume it is in the worse case scenario (it's not alone). In this case, LinK should not remove the IP directly. It should wait for another node to take the IP. To do so it will not renew the lease and wait for someone else to take it.
> 
> In the worse case this step can take 2*LeaseTime (1 for the lease to expire, 1 for another one to take it).

So with the change you propose, it won't be required to wait for this, as we can check directly if we're alone or not

**Comment by Soulou** [comment]
Right?

**Comment by johnsudaar** [comment]
Right. Plus with the watcher on the /ips prefix we could re-start an election right away so we'll never have to wait again.

**Comment by Soulou** [comment]
So beautiful :stars_in_the_eyes:

**Comment by Soulou** [comment]
I want it I want it now

**Comment by Soulou** [comment]
**Caprice**

**Comment by Soulou** [comment]
But btw, not sure it's "pretty straight forward and super easy"

**Comment by johnsudaar** [comment]
And we don't ask clients to manage stuff that should be manage internal.

**Comment by johnsudaar** [comment]
No it's not but meh.

**Comment by Soulou** [comment]
Yep I agree, it's a much nicer design, I did this PR because it was a problem for us, as I wanted to find ways to make our etcd cluster scale, and apply it quickly, I agree it's not a good method after this conversation. (it was maybe the quickest)

**Pull request merged**